### PR TITLE
Develop

### LINKED ISSUE #22: Error response data
For failed responses debugging would be useful to print to console the response data including headers. This usually include a request id and validation messages in case of 4XX exit codes.

### LINKED ISSUE #94: HAR as schema
This is a common format for http structures.
https://en.wikipedia.org/wiki/.har
http://www.softwareishard.com/blog/har-12-spec/#request

Google Chrome uses it to archive a request also.

This should be the schema used for the request object.  After implementation is complete, this issue can be implemented as a standard:

https://github.com/Huachao/vscode-restclient/issues/287

### LINKED ISSUE #97: Schema validation fails silently
**Describe the bug**
I've written a request that fails schema validation but the strest run does not fail.

**To Reproduce**
- create a request yml that is invalid
- run it
- the run passes

**Expected behavior**
- the run should fail with a non-zero exit code

**Additional context**

```
[ Strest ] Found 3 test file(s)
[ Strest ] Schema validation: 2 of 3 file(s) passed

Executing tests in ./
Executing tests in: /app/tests/
Executing tests in: /app/tests/tokens/
✔ Testing userToken succeeded (0.573s)

[ Strest ] ✨  Done in 0.634s
```

The only indication that something went wrong is `2 of 3 file(s) passed` at the top, which is easy to miss, especially in longer scripts.

### PR DESCRIPTION
**Pull request by jgroom33**
This is ready for review.  It implements HAR as the schema.

- [x] Update README(s) 
- [x] Add regex validation
```yaml
validate:
- jsonpath: status
  regex: 2\d+
```
- [x] saved responses stored as har
```yml
...
    request:
      url: https://postman-echo.com/get
      method: GET
      queryString:
      - name: foo
        value: bar
  value2:
    request:
      url: https://postman-echo.com/get
      method: GET
      queryString:
      - name: baz
        value: <$ value1.content.args.foo $>
```

Resolve #94 
Resolve #97 
Resolve #22

### CODE REVIEW

**Comment by jgroom33** [comment]
@shrikster could you help review this? I think it's ready for release, but would like you opinion.

**Comment by eykrehbein** [comment]
Reviewing tomorrow 👍 @jgroom33

File: src/test.ts [hunk 1]
```diff
@@ -1,30 +1,45 @@
...
  */
-let requestReponses: any = {
-  // // Example data
-  // register: {
-  //     id: 123,
-  //     token: 'aTokenValue'
-  // },
-  // rawDataExample: 'asdaasds'
-}
+let requestReponses: Map<string, object> = new Map<string, object>()
```
[the remark this answers is not in the document]
Comment (jgroom33) [reply src/test.ts]: I'm not sure that moving it inside would work. Currently, it's a global that each request adds to.
There's probably a more structured way to do this. Maybe with a class?
A refactor would probably be a good goal... Perhaps when the api feature is implemented.

File: src/handler.ts [hunk 2]
```diff
@@ -51,8 +51,10 @@ export const start = async (dir:string , cmd: any) => {
   writeMessage(`Schema validation: ${colorizedAmountOfValidSchemas} of ${colorizedTestFileAmount} file(s) passed`)
 
   const validTests = validateSchema.proofedSettings;
-  if(validTests.length === 0) {
-    console.log();
+  if(validateSchema.errors.length > 0) {
+    for (let erroredSchema of validateSchema.errors){
+      console.log(JSON.stringify(erroredSchema.details, null, 2));
```
[the remark this answers is not in the document]
Comment (jgroom33) [reply src/handler.ts]: I updated this to use writeErrorMessage()

File: src/test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jgroom33) [reply src/test.ts]: let's save this for a refactor... maybe something that uses classes and interfaces more.

File: src/test.ts [hunk 3]
```diff
@@ -34,57 +49,76 @@ let definedVariables: any = {
...
+        let bypass = false
+        if (cmd.key){
+          bypass = true
+          if (cmd.key == requestName) {
+            bypass = false
+          }
+        }
+
+        if (!abortBecauseTestFailed && !bypass) {
+          const val: requestsObjectSchema = requests[requestName];
```
[the remark this answers is not in the document]
Comment (jgroom33) [reply src/test.ts]: it's using this:

```javascript
export interface requestsObjectSchema {
  delay: number,
  maxRetries: number,
  if: object,
  request: requestObjectSchema,
  validate: Array<validateObjectSchema>,
  log: boolean | string,
  auth: authObjectSchema,
}
```

I think the naming of the interfaces and schemas in configSchema.ts might benefit from some updating.  This should occur in a refactor as well.

**Comment by jgroom33** [comment]
The raw object allows the file to be parsed by nunjucks as each new response object is saved.
Roughly:
1. Process raw with nunjucks
2. Parse yaml
3. Execute request and store response
4. Repeat

This isn't really the way nunjucks was designed to be used, but it seems to work out.

**Comment by jgroom33** [comment]
@eykrehbein any other topics to discuss?

**Pull request merged**